File: sabq app ios/sabq/Components/BreakingTickerBar.swift
import SwiftUI
import Combine

/// شريط الأخبار العاجلة — the editorially-curated breaking-news strip driven by
/// the dashboard's "Breaking Ticker" manager (`GET /api/breaking-ticker/active`).
///
/// Mirrors the web `BreakingNewsTicker`: a coral bar carrying a bolt + "عاجل"
/// pill and headlines that auto-rotate every 5 seconds. Tapping a headline opens
/// its linked article (by slug, via `ArticleSlugRoute`) or external URL. Pager
/// dots appear when more than one headline is queued.
///
/// Distinct from `HomeFeedView.breakingNewsSection`, which surfaces a single
/// `newsType == "breaking"` *article*. This strip is curated copy, not an
/// article card, so it takes precedence at the top of the feed when active.
struct BreakingTickerBar: View {
    let headlines: [APIBreakingHeadline]

    @State private var index = 0
    @Environment(\.openURL) private var openURL

    /// 5s cadence matches the web ticker's auto-advance.
    private let timer = Timer.publish(every: 5, on: .main, in: .common).autoconnect()

    private var current: APIBreakingHeadline? {
        guard !headlines.isEmpty else { return nil }
        return headlines[min(index, headlines.count - 1)]
    }

    var body: some View {
        Group {
            if let current {
                link(for: current) {
                    bar(current)
                }
            }
        }
        .onReceive(timer) { _ in
            guard headlines.count > 1 else { return }
            withAnimation(.easeInOut(duration: 0.28)) {
                index = (index + 1) % headlines.count
            }
        }
        // Reset to the first headline whenever the dashboard swaps the active
        // topic (the headline set changes) so we never index past the new array.
        .onChange(of: headlines.map(\.id)) { _, _ in
            index = 0
        }
    }

    /// Wraps the bar in the right tap target: in-app article route when a slug
    /// is present, an external link when only a URL is set, otherwise plain.
    @ViewBuilder
    private func link<Content: View>(
        for headline: APIBreakingHeadline,
        @ViewBuilder content: () -> Content
    ) -> some View {
        if let slug = headline.linkedArticleSlug, !slug.isEmpty {
            NavigationLink(value: ArticleSlugRoute(slug: slug)) { content() }
                .buttonStyle(.plain)
        } else if let ext = headline.externalUrl, let url = URL(string: ext) {
            Button { openURL(url) } label: { content() }
                .buttonStyle(.plain)
        } else {
            content()
        }
    }

    private func bar(_ headline: APIBreakingHeadline) -> some View {
        HStack(spacing: 12) {
            HStack(spacing: 5) {
                Image(systemName: "bolt.fill")
                    .font(.system(size: 11, weight: .bold))
                Text("عاجل")
                    .font(.system(size: 13, weight: .heavy))
            }
            .foregroundStyle(.white)
            .padding(.horizontal, 10)
            .padding(.vertical, 5)
            .background(Capsule().fill(Color.white.opacity(0.20)))

            Text(headline.headline)
                .font(.system(size: 14.5, weight: .semibold))
                .foregroundStyle(.white)
                .lineLimit(2)
                .multilineTextAlignment(.leading)
                .frame(maxWidth: .infinity, alignment: .leading)
                .id(headline.id)
                .transition(.opacity)

            if headlines.count > 1 {
                HStack(spacing: 4) {
                    ForEach(headlines.indices, id: \.self) { i in
                        Circle()
                            .fill(Color.white.opacity(i == index ? 1 : 0.4))
                            .frame(width: 5, height: 5)
                    }
                }
            } else {
                Image(systemName: "chevron.left")
                    .font(.system(size: 12, weight: .bold))
                    .foregroundStyle(.white.opacity(0.7))
            }
        }
        .padding(.horizontal, 14)
        .padding(.vertical, 12)
        .frame(maxWidth: .infinity, alignment: .leading)
        .background(
            RoundedRectangle(cornerRadius: SabqTheme.tileRadius, style: .continuous)
                .fill(
                    LinearGradient(
                        colors: [SabqTheme.coral, SabqTheme.coral.opacity(0.82)],
                        startPoint: .leading,
                        endPoint: .trailing
                    )
                )
        )
        .accessibilityElement(children: .combine)
        .accessibilityLabel("عاجل: \(headline.headline)")
    }
}

File: android-native/app/src/main/kotlin/com/sabq/smart/ui/components/BreakingTickerBar.kt
package com.sabq.smart.ui.components

import androidx.compose.animation.Crossfade
import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.filled.KeyboardArrowLeft
import androidx.compose.material.icons.filled.Bolt
import androidx.compose.material3.Icon
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableIntStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.sabq.smart.data.BreakingTickerHeadline
import com.sabq.smart.ui.theme.SabqTheme
import kotlinx.coroutines.delay

/**
 * شريط الأخبار العاجلة — the editorially-curated breaking-news strip
 * driven by the dashboard's "Breaking Ticker" manager
 * (`GET /api/breaking-ticker/active`). 1:1 port of iOS
 * `Components/BreakingTickerBar.swift`: a coral gradient bar carrying a
 * bolt + "عاجل" capsule and headlines that auto-rotate every 5 seconds.
 * Pager dots appear when more than one headline is queued; a chevron
 * otherwise.
 *
 * Distinct from the single breaking-article pill (HomeFeedScreen's
 * `BreakingNewsPill`), which is the fallback when no ticker topic is
 * active. Tap handling (article slug vs. external URL) is resolved by
 * the caller via [onHeadlineClick].
 */
@Composable
fun BreakingTickerBar(
    headlines: List<BreakingTickerHeadline>,
    onHeadlineClick: (BreakingTickerHeadline) -> Unit,
) {
    if (headlines.isEmpty()) return
    // Reset to the first headline whenever the dashboard swaps the
    // active topic (the headline set changes) so we never index past
    // the new array — mirrors iOS `.onChange(of: headlines.map(\.id))`.
    val idsKey = headlines.joinToString("|") { it.id }
    var index by remember(idsKey) { mutableIntStateOf(0) }

    // 5s cadence matches the web ticker's auto-advance (iOS Timer).
    LaunchedEffect(idsKey) {
        if (headlines.size > 1) {
            while (true) {
                delay(5_000)
                index = (index + 1) % headlines.size
            }
        }
    }

    val current = headlines[index.coerceIn(0, headlines.size - 1)]
    val coral = SabqTheme.colors.coral
    val shape = RoundedCornerShape(SabqTheme.dimens.tileRadius)

    Row(
        modifier = Modifier
            .fillMaxWidth()
            .clip(shape)
            .background(
                Brush.horizontalGradient(listOf(coral, coral.copy(alpha = 0.82f))),
                shape,
            )
            .clickable { onHeadlineClick(current) }
            .padding(horizontal = 14.dp, vertical = 12.dp),
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(12.dp),
    ) {
        // Bolt + "عاجل" capsule on white 20% — iOS `Capsule().fill(...)`.
        Row(
            modifier = Modifier
                .clip(CircleShape)
                .background(Color.White.copy(alpha = 0.20f), CircleShape)
                .padding(horizontal = 10.dp, vertical = 5.dp),
            verticalAlignment = Alignment.CenterVertically,
            horizontalArrangement = Arrangement.spacedBy(5.dp),
        ) {
            Icon(
                imageVector = Icons.Filled.Bolt,
                contentDescription = null,
                tint = Color.White,
                modifier = Modifier.size(12.dp),
            )
            Text(
                text = "عاجل",
                style = SabqTheme.typography.metaSmall.copy(
                    fontSize = 12.sp,
                    fontWeight = FontWeight.SemiBold,
                    color = Color.White,
                ),
            )
        }

        // Rotating headline — crossfade matches iOS `.transition(.opacity)`.
        Crossfade(
            targetState = current,
            label = "breaking-ticker-headline",
            modifier = Modifier.weight(1f),
        ) { headline ->
            Text(
                text = headline.headline,
                style = SabqTheme.typography.cardTitle.copy(
                    fontSize = 14.sp,
                    fontWeight = FontWeight.Medium,
                    color = Color.White,
                ),
                maxLines = 2,
            )
        }

        if (headlines.size > 1) {
            Row(horizontalArrangement = Arrangement.spacedBy(4.dp)) {
                headlines.indices.forEach { i ->
                    Box(
                        modifier = Modifier
                            .size(5.dp)
                            .clip(CircleShape)
                            .background(Color.White.copy(alpha = if (i == index) 1f else 0.4f)),
                    )
                }
            }
        } else {
            Icon(
                imageVector = Icons.AutoMirrored.Filled.KeyboardArrowLeft,
                contentDescription = null,
                tint = Color.White.copy(alpha = 0.7f),
                modifier = Modifier.size(14.dp),
            )
        }
    }
}
